homeassistant/components/systemmonitor/coordinator.py

```diff
@@ -0,0 +1,233 @@
+"""DataUpdateCoordinators for the System monitor integration."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from datetime import datetime
+import logging
+import os
+import socket
+from typing import TypeVar
+
+import psutil
+from psutil._common import sdiskusage, shwtemp, sswap
+from psutil._pslinux import svmem
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_component import DEFAULT_SCAN_INTERVAL
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+IO_COUNTER = {
+    "network_out": 0,
+    "network_in": 1,
+    "packets_out": 2,
+    "packets_in": 3,
+    "throughput_network_out": 0,
+    "throughput_network_in": 1,
+}
+IF_ADDRS_FAMILY = {"ipv4_address": socket.AF_INET, "ipv6_address": socket.AF_INET6}
+# There might be additional keys to be added for different
+# platforms / hardware combinations.
+# Taken from last version of "glances" integration before they moved to
+# a generic temperature sensor logic.
+# https://github.com/home-assistant/core/blob/5e15675593ba94a2c11f9f929cdad317e27ce190/homeassistant/components/glances/sensor.py#L199
+CPU_SENSOR_PREFIXES = [
+    "amdgpu 1",
+    "aml_thermal",
+    "Core 0",
+    "Core 1",
+    "CPU Temperature",
+    "CPU",
+    "cpu-thermal 1",
+    "cpu_thermal 1",
+    "exynos-therm 1",
+    "Package id 0",
+    "Physical id 0",
+    "radeon 1",
+    "soc-thermal 1",
+    "soc_thermal 1",
+    "Tctl",
+    "cpu0-thermal",
+    "cpu0_thermal",
+    "k10temp 1",
+]
+
+_dataT = TypeVar(
+    "_dataT",
+    bound=sdiskusage
+    | sswap
+    | svmem
+    | int
+    | float
+    | str
+    | tuple[float, float, float]
+    | datetime
+    | bool
+    | None,
+)
+
+
+class MonitorCoordinator(DataUpdateCoordinator[_dataT]):
+    """A System monitor Base Data Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, type_: str, argument: str) -> None:
```

Comment:
I think we should remove argument and type from the common class and just add argument to the disk coordinator.